### PR TITLE
cancel stale reduce only orders

### PR DESCRIPTION
**Pull request by atvanguard**
## Why this should be merged
Cancel stale reduce only orders

### CODE REVIEW

File: plugin/evm/orderbook/memory_database.go [hunk 1]
```diff
@@ -1076,16 +1094,26 @@ func (db *InMemoryDatabase) determineOrdersToCancel(addr common.Address, trader
...
 		}
-		ordersToCancel[addr] = append(ordersToCancel[addr], order)
-		foundCancellableOrders = true
-		orderNotional := big.NewInt(0).Abs(hu.Div1e18(hu.Mul(order.GetUnFilledBaseAssetQuantity(), order.Price))) // | size * current price |
-		marginReleased := hu.Div1e6(hu.Mul(orderNotional, db.configService.getMinAllowableMargin()))
-		_availableMargin.Add(_availableMargin, marginReleased)
-		if _availableMargin.Sign() >= 0 {
-			break
+		if order.ReduceOnly {
+			if marketsToCancelReduceOnlyOrdersIn[order.Market] {
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database.go:1102]: this line can give a panic cuz we don't set false in this map for any market. It should be written like:

```
if shouldCancel, _ := marketsToCancelReduceOnlyOrdersIn[order.Market]; shouldCancel {
```

File: plugin/evm/orderbook/memory_database.go [hunk 2]
```diff
@@ -1076,16 +1094,26 @@ func (db *InMemoryDatabase) determineOrdersToCancel(addr common.Address, trader
...
+				ordersToCancel[addr] = append(ordersToCancel[addr], order)
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database.go:1103]: `foundCancellableOrders` should also be set to true here

File: plugin/evm/orderbook/memory_database.go [hunk 3]
```diff
@@ -1076,16 +1094,26 @@ func (db *InMemoryDatabase) determineOrdersToCancel(addr common.Address, trader
...
+				marketsToCancelReduceOnlyOrdersIn[order.Market] = false // keeping it simple by cancelling just 1 stale order per market per run
+			} else {
+				continue
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database.go:1106]: this bit is not required, can be removed

**Review by shubhamgoyal42 — CHANGES_REQUESTED** [review]

File: plugin/evm/orderbook/memory_database.go [hunk 4]
```diff
@@ -1033,19 +1035,35 @@ func (db *InMemoryDatabase) GetNaughtyTraders(hState *hu.HubbleState) ([]Liquida
 			liquidablePositions = append(liquidablePositions, determinePositionToLiquidate(trader, addr, marginFraction, hState.ActiveMarkets, minSizes))
 			continue // we do not check for their open orders yet. Maybe liquidating them first will make available margin positive
 		}
+
+		marketsToCancelReduceOnlyOrdersIn := make(map[int]bool)
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply plugin/evm/orderbook/memory_database.go:1040]: Good point. Let's keep it this way for now and later look at refactoring this. There's some more changes required so we can club them together.

File: plugin/evm/orderbook/memory_database.go (hunk 3 above)
Comment (atvanguard) [reply plugin/evm/orderbook/memory_database.go]: done

File: plugin/evm/orderbook/memory_database.go [hunk 5]
```diff
@@ -1048,26 +1049,36 @@ func (db *InMemoryDatabase) GetNaughtyTraders(hState *hu.HubbleState) ([]Liquida
...
+			// negative available margin, and has reserved margin
+			if trader.Margin.Reserved.Sign() > 0 {
+				shouldLookForOrdersToCancel = true
+			} else {
+				// this has the affect that we will not look for non-reduce only orders to cancel
+				availableMargin = big.NewInt(0)
+			}
+		}
+
+		if !shouldLookForOrdersToCancel {
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database.go:1068]: nice!

**Review by shubhamgoyal42 — APPROVED** [review]

**Pull request merged**